### PR TITLE
[FE] S3 onboarding with healthchecks

### LINKED ISSUE #2523: Web: Show healthcheck errors after an s3 source is saved
### Description

For s3 sources, we are going to allow them to be saved even if the healthcheck fails. However, in the final step of the onboarding or the update forms, we would like to show the user any healthcheck issues present. 
If healthcheck is green, we just show the existing final page with the green check mark.

See https://github.com/panther-labs/panther/issues/2522 for more info about the healthcheck errors page.

### Acceptance Criteria

- If after an **s3** source is saved the healthcheck is red, display the healthcheck errors to the user. 
- This page will have 2 buttons. One for the user to retry the healthcheck and one to navigate to the form and edit the source. This would redirect the user to update the source, as the source would be already saved in the db.

### PR DESCRIPTION
**Pull request by alexmylonas**
## Description

With the latest changes, we allow users to onboard an S3 Log source even if 1 or many healthchecks fail.

This PR addresses the changes required for users to understand that they have successfully onboarded a S3 Log source but they have some warning that make the integration un-usable.

Closes #2523 

## Designs

Mocks [here](https://app.abstract.com/projects/558a8de6-7134-4c8b-91c5-62074bb1279b/branches/df9d59c8-eeb1-452d-a5f5-3712b64010be/commits/0259dc1138c4d66d3db2d1da2b8506bbfc83818d/files/3f113d4e-19e8-43b1-8e2f-148ac5f988c8/layers/4AD411B2-C8BC-4CFB-A1FD-8A4DD39AC708?mode=build&selected=root-81492A88-53FB-4BE5-9B5B-513832B16DE3&sha=0259dc1138c4d66d3db2d1da2b8506bbfc83818d)

## Screencast

https://user-images.githubusercontent.com/14179917/107762910-64b24b80-6d36-11eb-94f1-7fc9029b3431.mov

### Checklist:

- [x] Make sure you give a high-level overview of the problem & the solution
- [x] Make sure the related issue is referenced
- [x] Make sure screenshots are added if the UI is changed in any way
- [x] Make sure all of the significant logic is covered by tests
- [x] Make sure your changes are accessible (addition of aria-attributes & correct HTML semantics)
- [x] Make sure you reference the related design document if it exists
- [x] Make sure telemetry is added (analytics & errors)
- [x] ~~Make sure the changes are responsive (appearing decently on devices)~~

### CODE REVIEW

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+interface SuccessContentProps {
+  integrationId: string;
+}
+
+const SuccessContent: React.FC<SuccessContentProps> = ({ integrationId }) => {
+  const { reset: resetWizard } = useWizardContext();
+  const { initialValues, resetForm } = useFormikContext<S3LogSourceWizardValues>();
+
+  const [getS3, { data, loading }] = useGetS3LogSourceLazyQuery({
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx:39]: can be a simple query. Just use `refetch`
```suggestion
  const { data, loading, refetch } = useGetS3LogSource({
```

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    fetchPolicy: 'network-only', // Don't use cache
+    variables: { id: integrationId },
+  });
+
+  React.useEffect(() => {
+    getS3();
+  }, []);
+
+  if (loading || !data) {
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx:48]:
```suggestion
  if (loading) {
```

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    healthChecks.push(getObjectStatus);
+  }
+  const isHealthy = healthChecks.every(healthMetric => Boolean(healthMetric.healthy));
+
+  if (!isHealthy) {
+    return (
+      <WizardPanel>
+        <Flex align="center" direction="column" mx="auto" width={675}>
+          <WizardPanel.Heading
+            title="The source is saved, but there are some issues blocking log integration"
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx:74]:
```suggestion
            title="The source is saved, but there are some issues blocking log injestion"
```

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
+/**
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx:1]: rename to `HealthPanel`?

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx [hunk 5]
```diff
@@ -51,20 +49,25 @@ function getResponseData(result: SubmitResult): S3LogIntegrationDetails {
...
     (async () => {
       try {
         const result = ((await submitForm()) as unknown) as SubmitResult;
-        const { managedBucketNotifications, notificationsConfigurationSucceeded } = getResponseData(
-          result
-        );
+        const {
+          managedBucketNotifications,
+          notificationsConfigurationSucceeded,
+          integrationId: newIntegrationId,
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:68]: should we use `setWizardData` here?

**Review by 3nvi — COMMENTED** [review]
@alexmylonas amazing work. Couple of suggestions and we are g2g

**Comment by giorgosp** [comment]
Cool! The behavior is the same during updates too, right?

**Comment by kostaspap** [comment]
We want users to be able to complete onboarding even if the health check fails. In the video it seems that the only options are "edit"  or "retry" health check.

**Comment by kostaspap** [comment]
> We want users to be able to complete onboarding even if the health check fails. In the video it seems that the only options are "edit" or "retry" health check. So we probably want a 3rd button in the last screen that allows the users to acknowledge the issue and just complete the process. @giorgosp
> ~Currently, it is allowed to just onboard an unhealthy source. The whole idea of this change is to just provide an additional warning to the users, but still let them complete the process.~

I see that it says in the top "The source is saved", just thinking if it would better UX if the UI was somehow allowing users to acknowledge issue and exit the wizard. Someone might find it confusing otherwise?

**Comment by alexmylonas** [comment]
> Cool! The behavior is the same during updates too, right?

Yeap, same behaviour.

> We want users to be able to complete onboarding even if the health check fails. In the video it seems that the only options are "edit" or "retry" health check. So we probably want a 3rd button in the last screen that allows the users to acknowledge the issue and just complete the process. @giorgosp
> ~Currently, it is allowed to just onboard an unhealthy source. The whole idea of this change is to just provide an additional warning to the users, but still let them complete the process.~

That sounds valid, should we inform @dimitriszoitas to update design?

**Comment by kostaspap** [comment]
> > Cool! The behavior is the same during updates too, right?
> 
> Yeap, same behaviour.
> 
> > We want users to be able to complete onboarding even if the health check fails. In the video it seems that the only options are "edit" or "retry" health check. So we probably want a 3rd button in the last screen that allows the users to acknowledge the issue and just complete the process. @giorgosp
> > ~Currently, it is allowed to just onboard an unhealthy source. The whole idea of this change is to just provide an additional warning to the users, but still let them complete the process.~
> 
> That sounds valid, should we inform @dimitriszoitas to update design?

Yup, I've put a relevant comment in his designs. 
Note that personally I don't think it is necessary to block this PR for the proposed change. I'll leave it up to you if you want to make the change as part of this PR or in a subsequent PR.

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx (hunk 5 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:71]: Yeap moved this to `WizardContext`

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx (hunk 4 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx]: Renamed!

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx (hunk 3 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx]: Yeap changed

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx (hunk 2 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx]: Changed!

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx (hunk 1 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/SuccessContent.tsx]: Good call! Done.

File: web/src/assets/icons/icon-filled-white.svg [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (alexmylonas) [reply web/src/assets/icons/icon-filled-white.svg]: renamed to warning

File: web/src/components/Wizard/Wizard.tsx [hunk 7]
```diff
@@ -118,18 +119,22 @@ function Wizard<WizardData = any>({
       reset: resetWizard,
       data: wizardData,
       currentStepStatus,
+      integrationId,
```
Comment (3nvi) [inline web/src/components/Wizard/Wizard.tsx:122]: all this doesn't belong in the Wizard. It's domain specific while the wizard is & should remain domain-agnostic

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx (hunk 5 above)
Comment (3nvi) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:71]: I don't think the implementation aligns well with what I had in mind. We shouldn't let the `Wizard` component know about our problem here at all. It should be agnostic of that.

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import urls from 'Source/urls';
+import React from 'react';
+import { Link as RRLink } from 'react-router-dom';
+import { useFormikContext } from 'formik';
+import { S3LogSourceWizardValues } from 'Components/wizards/S3LogSourceWizard/S3LogSourceWizard';
+import { useGetS3LogSource } from 'Pages/EditS3LogSource/graphql/getS3LogSource.generated';
+import HealthCheckWarning from './HealthCheckWarning';
+
+const HealthCheckPanel: React.FC = () => {
+  const { reset: resetWizard, data: wizardData } = useWizardContext();
```
Comment (alexmylonas) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:32]: take `integrationId` from `wizardData`

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx [hunk 9]
```diff
@@ -62,9 +65,12 @@ const ValidationPanel: React.FC = () => {
...
+        } = getResponseData(result);
+        updateData({ integrationId: newIntegrationId });
```
Comment (alexmylonas) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:73]: store `integrationId` on `wizardData`

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx (hunk 5 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:71]: Gotcha, stored value on `wizardData` instead

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx (hunk 8 above)
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:32]: can we add a typing here?

```suggestion
  const { reset: resetWizard, data: wizardData } = useWizardContext<{ integrationId?: string }>();
```

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  const { initialValues, resetForm } = useFormikContext<S3LogSourceWizardValues>();
+
+  const { data, loading, refetch } = useGetS3LogSource({
+    notifyOnNetworkStatusChange: true,
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:36]: do we need that?

```suggestion

```

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    getObjectStatus,
+    kmsKeyStatus,
+    s3BucketStatus,
+  } = data?.getS3LogIntegration.health;
+
+  const healthChecks = [processingRoleStatus, kmsKeyStatus, s3BucketStatus];
+  if (getObjectStatus) {
+    healthChecks.push(getObjectStatus);
+  }
+  const isHealthy = healthChecks.every(healthMetric => Boolean(healthMetric.healthy));
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:60]: can we not do:

`const isHealthy = Object.values(data?.getS3LogIntegration.health).filter(Boolean).every(metric => metric.healthy)`?

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx [hunk 12]
```diff
@@ -50,8 +48,13 @@ function getResponseData(result: SubmitResult): S3LogIntegrationDetails {
...
 const ValidationPanel: React.FC = () => {
   const [errorMessage, setErrorMessage] = React.useState('');
-  const { reset: resetWizard, currentStepStatus, setCurrentStepStatus } = useWizardContext();
-  const { initialValues, submitForm, resetForm } = useFormikContext<S3LogSourceWizardValues>();
+  const {
+    reset: resetWizard,
+    currentStepStatus,
+    setCurrentStepStatus,
+    updateData,
+  } = useWizardContext();
```
Comment (3nvi) [inline web/src/components/wizards/S3LogSourceWizard/ValidationPanel/ValidationPanel.tsx:56]:
```suggestion
  } = useWizardContext<{ integrationId?: string }>();
```

**Review by 3nvi — APPROVED** [review]
Looks good! A couple of suggestions.

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx (hunk 11 above)
Comment (giorgosp) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:60]: +1, I have added another healthcheck in #2684 and this snippet will handle it nicely without extra code in either PRs

Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx]: Good idea, implemented

File: web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx (hunk 10 above)
Comment (alexmylonas) [reply web/src/components/wizards/S3LogSourceWizard/ValidationPanel/HealthCheckPanel.tsx:36]: Without loading doesn't turn `true` when `refetch` is triggered

**Pull request merged**